File: cpp/algorithms/advanced/mo_algorithm/mo_algorithm.cpp
// 莫队：区间不同元素个数（与 Python 版同构）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

vector<int> mo_distinct_count(const vector<int>& arr, const vector<pair<int, int>>& queries) {
    const int n = static_cast<int>(arr.size());
    const int qn = static_cast<int>(queries.size());
    if (qn == 0) return {};
    const int b = max(1, static_cast<int>(sqrt(n)));
    vector<int> order(qn);
    iota(order.begin(), order.end(), 0);
    sort(order.begin(), order.end(), [&](int i, int j) {
        int li = queries[i].first / b, lj = queries[j].first / b;
        if (li != lj) return li < lj;
        if (li & 1) return queries[i].second > queries[j].second;
        return queries[i].second < queries[j].second;
    });
    unordered_map<int, int> cnt;
    int cur = 0;
    vector<int> out(qn);
    auto add = [&](int i) {
        int x = arr[i];
        if (++cnt[x] == 1) ++cur;
    };
    auto remove = [&](int i) {
        int x = arr[i];
        if (--cnt[x] == 0) --cur;
    };
    int L = 0, R = -1;
    for (int qi : order) {
        int l = queries[qi].first, r = queries[qi].second;
        while (R < r) add(++R);
        while (R > r) remove(R--);
        while (L < l) remove(L++);
        while (L > l) add(--L);
        out[qi] = cur;
    }
    return out;
}

int main() {
    vector<int> a{1, 2, 1, 3, 2, 3, 3};
    vector<pair<int, int>> qs{{0, 6}, {1, 4}, {4, 4}};
    auto ans = mo_distinct_count(a, qs);
    assert(ans.size() == 3 && ans[0] == 3 && ans[1] == 3 && ans[2] == 1);
    cout << "mo_algorithm OK" << endl;
    return 0;
}

File: python/algorithms/advanced/mo_algorithm/mo_algorithm.py
"""莫队（离线）：区间不同元素个数。

将询问按左端点所在块排序，块内右端点交替升降序，均摊 O((n+q)·√n) 量级移动指针。
"""

from __future__ import annotations

from collections import defaultdict
from typing import List, Tuple


def mo_distinct_count(arr: List[int], queries: List[Tuple[int, int]]) -> List[int]:
    """queries 为闭区间 [l, r]；返回与 queries 同序的答案列表。"""
    n = len(arr)
    qn = len(queries)
    if qn == 0:
        return []
    b = max(1, int(n**0.5))

    def sort_key(i: int) -> Tuple[int, int]:
        l, r = queries[i]
        bl = l // b
        if bl & 1:
            return (bl, -r)
        return (bl, r)

    order = sorted(range(qn), key=sort_key)
    cnt: defaultdict[int, int] = defaultdict(int)
    cur = 0
    out = [0] * qn

    def add(i: int) -> None:
        nonlocal cur
        x = arr[i]
        cnt[x] += 1
        if cnt[x] == 1:
            cur += 1

    def remove(i: int) -> None:
        nonlocal cur
        x = arr[i]
        cnt[x] -= 1
        if cnt[x] == 0:
            cur -= 1

    L, R = 0, -1
    for qi in order:
        l, r = queries[qi]
        while R < r:
            R += 1
            add(R)
        while R > r:
            remove(R)
            R -= 1
        while L < l:
            remove(L)
            L += 1
        while L > l:
            L -= 1
            add(L)
        out[qi] = cur
    return out


if __name__ == "__main__":
    a = [1, 2, 1, 3, 2, 3, 3]
    assert mo_distinct_count(a, [(0, 6), (1, 4), (4, 4)]) == [3, 3, 1]
    print("mo_algorithm OK")
